transform/transformers/transformer.py

```diff
@@ -0,0 +1,156 @@
+import io
+import logging
+import os
+import tempfile
+import zipfile
+from collections import OrderedDict
+
+from reportlab.lib.pagesizes import A4
+from reportlab.platypus import SimpleDocTemplate
+from transform.transformers.cs_formatter import CSFormatter
+from transform.transformers.survey import Survey
+from transform.transformers.ImageTransformer import ImageTransformer
+from transform.transformers.PDFTransformer import PDFTransformer
+from structlog import wrap_logger
+
+
+__doc__ = "Transform survey data into formats required downstream."
+
+
+class Transformer:
+
+    """A base class for SDX transformers.
+
+    Subclasses should define the contents of the following class variables:
+
+        * :py:const:`defn`
+        * :py:const:`package`
+        * :py:const:`pattern`
+
+    """
+    #: Transformer subclasses declare their transforms in this class variable.
+    #: Each element is a 3-tuple consisting of:
+    #:
+    #: #. An integer or range corresponding to one or more question ids.
+    #: #. A default value for the question(s).
+    #: #. A :ref:`Processing function <processors>`.
+    #:
+    #: Eg, to declare question ids 151, 152, 153 as unsigned integers with a default
+    #: of 0::
+    #:
+    #:  defn = [
+    #:   (range(151, 154, 1), 0, Processor.unsigned_integer),
+    #:
+    #:  ]
+    #:
+    defn = []
+
+    #: Defines the package where survey definitions can be found.
+    #: For deployed Python packages, this will be a dotted package name.
+    #: If your Transformer class is deployed from a source repository,
+    #: you should set this class variable to `__name__`.
+    package = "sdx.common"
+
+    #: Defines a search pattern for survey definitions based on identifiers.
+    #: The path is relative to the location specified by :py:const:`package` above.
+    pattern = "../surveys/{survey_id}.{inst_id}.json"
+
+    @classmethod
+    def ops(cls):
+        """Publish the sequence of operations for the transform.
+
+        Return an ordered mapping from question id to default value and processing function.
+
+        """
+        return OrderedDict([
+            ("{0:02}".format(qNr), (dflt, fn))
```

Comment:
These variable names are pretty terrible at a glance.  Doesn't need to be fixed for this PR but might be worth noting somewhere to get them improved

Reply:
Agreed - the MWSS transformer benefited significantly from some extra docs and more verbose variable names. A two character name is actually a violation of PyLint.